### PR TITLE
Add both variants of check4 in TestSQLInjection

### PR DESCRIPTION
**Pull request by rnehra01**
This check is an extension of the original **check(4)**.
 It uses **multiplication** variant instead.
Also, repetition of codes of both checks(**mult** and **plus**) has been addressed by extracting a new method.

### CODE REVIEW

**Comment by kingthorin** [comment]
Thanks

**Comment by rnehra01** [comment]
When will this pull be merged? I have to submit other changes.

**Comment by thc202** [comment]
Pull requests are usually merged once reviewed and approved by more than one person.

Which other changes do you want to submit?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 1]
```diff
@@ -779,6 +777,103 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+										sendAndReceive(msg4aConfirm, false); //do not follow redirects
+									} catch (SocketException ex) {
+										if (log.isDebugEnabled()) log.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() + 
+												" when accessing: " + msg4aConfirm.getRequestHeader().getURI().toString());
+										continue; //Something went wrong, continue to the next item in the loop
+									}
+									countExpressionBasedRequests++;
+
+									String confirmExpressionOutputStripped4a = msg4aConfirm.getResponseBody().toString();
+									String confirmExpressionOutputUnstripped4a = this.stripOff(confirmExpressionOutputStripped4a, modifiedParamValueConfirm4a);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The variables names were swapped (which is misleading), was that on purpose?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 2]
```diff
@@ -779,6 +777,103 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+
+									//set up two little arrays to ease the work of checking the unstripped output or the stripped output
+									String confirmExpressionBodyOutput4a[] = {confirmExpressionOutputUnstripped4a, confirmExpressionOutputUnstripped4a};
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Stripped variable was replaced with "unstripped", was that on purpose?

**Comment by rnehra01** [comment]
@thc202, Requested changes has been done.
Code repetitions is addressed using a **loop**

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 3]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
 					if (this.debugEnabled) {
 						log.debug("The parameter value [" + origParamValue + "] is of type Integer");
 					}
+					// This check is implemented using two variant PLUS(+) and MULT(*)
+					// PLUS variant check the param value "3-2" gives same result and param value "4-2" gives different result if original param value is 1
+					// PLUS variant check the param value "2/2" gives same result and param value "4/2" gives different result if original param value is 1
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: They're not both PLUS variants.

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: It might help if the text is slightly more specific "...gives the same result as the original request..."

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 4]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+					String check4Variant []= {"plus", "mult"};
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: I'd suggest going with caps like you did in the comments.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 5]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
-					countExpressionBasedRequests++;
-
-					String modifiedExpressionOutputUnstripped = msg4.getResponseBody().toString();
-					String modifiedExpressionOutputStripped = this.stripOff(modifiedExpressionOutputUnstripped, modifiedParamValue);
-
-					//set up two little arrays to ease the work of checking the unstripped output, and then the stripped output
-					String normalBodyOutput[] = {mResBodyNormalUnstripped, mResBodyNormalStripped};
-					String expressionBodyOutput[] = {modifiedExpressionOutputUnstripped, modifiedExpressionOutputStripped};
-					boolean strippedOutput[] = {false, true};
+					for (int booleancheck4VariantIndex = 0; booleancheck4VariantIndex<2 && !sqlInjectionFoundForUrl && countExpressionBasedRequests < doExpressionMaxRequests; booleancheck4VariantIndex++) {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The c in check should be capitalized `booleanCheck4VariantIndex `.  (Is the leading "boolean" necessary? If so shouldn't it actually be "expression"?)
Also probably best to check against `check4Variant.length` instead of statically `2` (make maintenance more friendly in the future if another variant is added).

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 6]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+									String modifiedParamValueConfirm = "";
+
+									if (check4Variant[booleancheck4VariantIndex] == "plus"){
+										//and set the parameter value to a string value like "4-2", if the original parameter value was "1"
+										int paramPlusTwo = paramAsInt + 3;
+										modifiedParamValueConfirm = String.valueOf(paramPlusTwo) + "-2";
+									}
+									else if (check4Variant[booleancheck4VariantIndex] == "mult"){
+										//and set the parameter value to a string value like "4/2", if the original parameter value was "1"
+										int paramPlusTwo = paramAsInt * 4;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The variable name should change, it isn't a "PlusTwo" at this point. Perhaps `increasedParamValueConfirm` or `modifiedParamValueConfirm` or something more generic like that to fit both variants?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 7]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
-							int paramPlusFour = paramAsInt + 3;
-							String modifiedParamValueConfirm = String.valueOf(paramPlusFour) + "-2";
-
-							//and prepare a request to set the parameter value to a string value like "4-2", if the original parameter value was "1"
-							//Note that the two values are NOT equivalent, and the param value is different to the original
-							HttpMessage msg4Confirm = getNewMsg();
-							setParameter(msg4Confirm, param, modifiedParamValueConfirm);
+							else if (check4Variant[booleancheck4VariantIndex] == "mult"){
+								//and set the parameter value to a string value like "2/2", if the original parameter value was "1"
+								int paramPlusTwo = paramAsInt * 2;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The variable name should change, it isn't a "PlusTwo" at this point. Perhaps `increasedParamValue` or `modifiedParamValue` or something more generic like that to fit both variants?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 8]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+										log.debug("Check 4, " + (strippedOutput[booleanStrippedUnstrippedIndex] ? "STRIPPED" : "UNSTRIPPED") + " html output for modified expression parameter [" + modifiedParamValue + "] matched (refreshed) original results for " + refreshedmessage.getRequestHeader().getURI());
+									}
+									//confirm that a different parameter value generates different output, to minimise false positives
+									// this time get a bigger value
+
+									String modifiedParamValueConfirm = "";
+
+									if (check4Variant[booleancheck4VariantIndex] == "plus"){
+										//and set the parameter value to a string value like "4-2", if the original parameter value was "1"
+										int paramPlusTwo = paramAsInt + 3;
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: It isn't "PlusTwo" at this point.

**Comment by kingthorin** [comment]
Please do an interactive rebase and fix-up or squash into a single "atomic" commit: https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#what-we-zap-team-expect-from-you .

- https://www.kernel.org/pub/software/scm/git/docs/git-rebase.html#_interactive_mode
- https://wiki.eclipse.org/EGit/User_Guide#Interactive_Rebase

**Comment by rnehra01** [comment]
@kingthorin, do you want to make the two commits as one? But I put them separately because they have different purposes. One is related to adding a new feature and other fixing repetition.

**Comment by kingthorin** [comment]
@rnehra01 when it's merged it's all the same thing, adding the 4a variant so it should be a single commit from a project perspective. The history of your feature branch prior to being merged isn't too big a deal.

If you squash it then the commit message can be updated to clarify that the original 4 (plus) variant was added into a loop to reduce code repetition....

**Comment by rnehra01** [comment]
I hope this won't disappoint you.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 9]
```diff
@@ -692,93 +690,124 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
-										extraInfo, getSolution(), "", msg4);
-
-								sqlInjectionFoundForUrl = true;
 							}
-						}
-						//bale out if we were asked nicely
-						if (isStop()) { 
-							log.debug("Stopping the scan due to a user request");
-							return;
+						} catch (Exception e){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: I believe this is to handle the case mentioned by @kingthorin about the overflow when adding/multiplying the int? If so, this needs changes since no exception is thrown in that case.

**Comment by thc202** [comment]
Not a blocker for me (the whole scan method is in need of refactoring anyway) but I still think it would be better to extract a method, it's easier to read/change:
```Java
expressionBasedAttack(paramAsInt + 2 + "-2", paramAsInt + 3 + "-2");
expressionBasedAttack(paramAsInt * 2 + "/2", paramAsInt * 4 + "/2");
```
(The expressions could also be extracted to variables to improve readability, the good thing is that they are defined and used in the same place.)

than the current way.

**Comment by kingthorin** [comment]
Build is failing now. Did you leverage something Java 8 specific?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 10]
```diff
@@ -692,93 +690,127 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
-									extraInfo = Constant.messages.getString(MESSAGE_PREFIX + "alert.expressionbased.extrainfo", modifiedParamValue, "NOT ");
+							String modifiedExpressionOutputUnstripped = msg4.getResponseBody().toString();
+							String modifiedExpressionOutputStripped = this.stripOff(modifiedExpressionOutputUnstripped, modifiedParamValue);
+
+							//set up two little arrays to ease the work of checking the unstripped output, and then the stripped output
+							String normalBodyOutput[] = {mResBodyNormalUnstripped, mResBodyNormalStripped};
+							String expressionBodyOutput[] = {modifiedExpressionOutputUnstripped, modifiedExpressionOutputStripped};
+							boolean strippedOutput[] = {false, true};
+
+							for (int booleanStrippedUnstrippedIndex = 0; booleanStrippedUnstrippedIndex < 2 && !sqlInjectionFoundForUrl && countExpressionBasedRequests < doExpressionMaxRequests; booleanStrippedUnstrippedIndex++) {
```
Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: @thc, if I create the method you suggested then **countExpressionBasedRequests** will need to be passed as the parameter in that method and method has to return two things **sqlInjectionForUrl(boolean)** and new **countExpressionBasedRequests**. Method returning two variables would be a mess in JAVA.
If we can remove the loop, then it would be easy. And I don't think any usage of comparing both stripped as well stripped. I think comparing stripped will do the same job.

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: `sqlInjectionFoundForUrl` could become a class variable instead of a method variable?

Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Yeah, some of the variables should be instance variables (like `doExpressionMaxRequests` is).

Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Regarding the comparisons, that's another subject that should be discussed separately (from these changes/pull request).

Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: I think a method `expressionBasedAttack` for checking **SQLI** will look odd it returns increased count of `doExpressionMaxRequests` instead of a **boolean** stating the presence of **SQLI**. But if you insist I will do it or I can do my proposed solution of removing **loop** in next PR.

Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Agreed, that would look odd, but no one is suggesting that. The method does not need to return any value if `sqlInjectionFoundForUrl` (`sqlInjectionAttack` and `countExpressionBasedRequests`) are instance variables.

Where's the proposed solution to remove the loop?

Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The loop could be removed by using only **stripped** comparison as there is no need for **unstripped** one.

Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Yeah, making both a **instance** variable will do the job.

Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The loop thing is if I only compare **stripped**, then **loop** can be removed.
Then make a method `expressionBasedAttack(paramAsInt + 2 + "-2")` that does not requires `countExpressionBasedRequests` as only one request is made so `countExpressionBasedRequests` can be incremented after method call.
But then we need to call it two times `expressionBasedAttack(paramAsInt + 2 + "-2")` and `expressionBasedAttack(paramAsInt + 3 + "-2")`

**Comment by rnehra01** [comment]
@thc202, I think this is what you wanted. 
I'll squash the commits once approved.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 11]
```diff
@@ -1451,6 +1490,42 @@ public static String getURLDecode(String msg) {
 		return result;
 	}
 
+	/**
+	* add two numbers with Arithmetic Overflow check
+	* @param firstNumber
+	* @param secondNumber
+	* @return
+	*/
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: This blank line should be removed

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 12]
```diff
@@ -1451,6 +1490,42 @@ public static String getURLDecode(String msg) {
...
+	    return (int) result;
+	}
+
+	/**
+	* multiply two numbers with Arithmetic Overflow check
+	* @param firstNumber
+	* @param secondNumber
+	* @return
+	*/
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: This blank line should be removed

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 13]
```diff
@@ -1451,6 +1490,42 @@ public static String getURLDecode(String msg) {
...
+	* add two numbers with Arithmetic Overflow check
+	* @param firstNumber
+	* @param secondNumber
+	* @return
+	*/
+
+	public static int addWithOverflowCheck(int firstNumber, int secondNumber) {
+	    long result = ((long) firstNumber) + ((long) secondNumber);
+	    if (result > Integer.MAX_VALUE) {
+	         throw new ArithmeticException("Overflow occured");
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Two r's in `occurred` ... applies to all 4 occurrences.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 14]
```diff
@@ -679,7 +692,7 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
 			//String mResBodyNormal = getBaseMsg().getResponseBody().toString();
 			mResBodyNormalUnstripped = refreshedmessage.getResponseBody().toString();
 			mResBodyNormalStripped = this.stripOff(mResBodyNormalUnstripped, origParamValue);
-
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Formatting changes should be separated into another PR. https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 15]
```diff
@@ -692,95 +705,42 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+						int paramMultFour = multiplyWithOverflowCheck(paramAsInt, 4);
+						String modifiedParamValueConfirmForMult = String.valueOf(paramMultFour) + "/2";
+						//Do the attack for MULT variant
+						expressionBasedAttack(param, modifiedParamValueForMult, modifiedParamValueConfirmForMult);
 						//bale out if we were asked nicely
 						if (isStop()) { 
 							log.debug("Stopping the scan due to a user request");
 							return;
 						}
+					} catch (Exception e){
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: You should catch the exceptions that are being thrown, that way if something else occurs it'll be more obvious `catch (ArithmeticException|RuntimeException ex)`

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 16]
```diff
@@ -692,95 +705,42 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+						if (this.debugEnabled) {
+							log.debug("The parameter value [" + origParamValue + "] is TOO large to test this check");
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Since you also caught underrun (thanks!) this message no longer works generically, perhaps:
`log.debug(""Caught " + ex.getClass().getName() + " " + ex.getMessage() + "When performing integer math with the parameter value [" + origParamValue + "]");`

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java (hunk 14 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Formatting changes should be submitted as a separate PR from code changes. https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#formatting

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java (hunk 11 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: This blank should be removed

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java (hunk 12 above)
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: This blank should be removed

**Comment by rnehra01** [comment]
@kingthorin, requested changes has been applied.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 17]
```diff
@@ -691,128 +705,41 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+						int paramMultFour = multiplyWithOverflowCheck(paramAsInt, 4);
+						String modifiedParamValueConfirmForMult = String.valueOf(paramMultFour) + "/2";
+						//Do the attack for MULT variant
+						expressionBasedAttack(param, modifiedParamValueForMult, modifiedParamValueConfirmForMult);
+						//bale out if we were asked nicely
+						if (isStop()) { 
+							log.debug("Stopping the scan due to a user request");
+							return;
+						}
+					} catch (ArithmeticException ex){
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The underflow condition is thrown as `RuntimeException` (unless that was just an oversight? L1503/1519) which isn't caught ... `catch (ArithmeticException|RuntimeException ex){`

Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: I think making that underflow throw `Arithmetic error` would be good.

**Comment by kingthorin** [comment]
Looks good to me. You can wait for further feedback from @thc202, don't forget that it'll need to be squashed.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 18]
```diff
@@ -75,7 +73,14 @@
 	 * Prefix for internationalised messages used by this rule
 	 */
 	private static final String MESSAGE_PREFIX = "ascanrules.testsqlinjection.";
-
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: This line could be kept.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 19]
```diff
@@ -75,7 +73,14 @@
 	 * Prefix for internationalised messages used by this rule
 	 */
 	private static final String MESSAGE_PREFIX = "ascanrules.testsqlinjection.";
-	
+	/**
+	* Did SQLInjection get found yet?
+	*/
+	private boolean sqlInjectionFoundForUrl = false;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The initialisation to false can be removed, that's the default value.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 20]
```diff
@@ -75,7 +73,14 @@
...
+	private String sqlInjectionAttack = null;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Same here (with `null`).

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 21]
```diff
@@ -92,6 +97,13 @@
 	private int doExpressionMaxRequests = 0;
 	private int doOrderByMaxRequests = 0;
 	//private int doStackedMaxRequests = 0;	//TODO: use in the stacked based implementation
+	//how many requests have we fired up?
+	private int countErrorBasedRequests = 0;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Same here (with zero).

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 22]
```diff
@@ -552,19 +564,20 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
-			int countExpressionBasedRequests = 0;
-			int countBooleanBasedRequests = 0;
-			int countUnionBasedRequests = 0;
-			int countOrderByBasedRequests = 0;
+			countErrorBasedRequests = 0;
+			countExpressionBasedRequests = 0;
+			countBooleanBasedRequests = 0;
+			countUnionBasedRequests = 0;
+			countOrderByBasedRequests = 0;
 			//int countStackedBasedRequests = 0;  //TODO: use in the stacked based queries implementation
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: This should be removed?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 23]
```diff
@@ -679,7 +692,6 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
 			//String mResBodyNormal = getBaseMsg().getResponseBody().toString();
 			mResBodyNormalUnstripped = refreshedmessage.getResponseBody().toString();
 			mResBodyNormalStripped = this.stripOff(mResBodyNormalUnstripped, origParamValue);
-
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The line should be restored.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 24]
```diff
@@ -692,95 +704,42 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+						//Do the attack for ADD variant
+						expressionBasedAttack(param, modifiedParamValueForAdd, modifiedParamValueConfirmForAdd);
+						//bale out if we were asked nicely
+						if (isStop()) { 
+							log.debug("Stopping the scan due to a user request");
+							return;
 						}
+						// MULT variant check the param value "2/2" gives same result as original request and param value "4/2" gives different result if original param value is 1
+						//set the parameter value to a string value like "2/2", if the original parameter value was "1"
+						int paramMultTwo = multiplyWithOverflowCheck(paramAsInt, 2);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Shouldn't do this attack if the SQLi was already found (or if reached `doExpressionMaxRequests`).

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 25]
```diff
@@ -692,95 +704,42 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
+						expressionBasedAttack(param, modifiedParamValueForMult, modifiedParamValueConfirmForMult);
 						//bale out if we were asked nicely
 						if (isStop()) { 
 							log.debug("Stopping the scan due to a user request");
 							return;
 						}
+					} catch (ArithmeticException ex){
+						if (this.debugEnabled) {
+							log.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() + "When performing integer math with the parameter value [" + origParamValue + "]");
+						}
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: There's a tab at the end of the line that should be removed.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 26]
```diff
@@ -692,95 +704,42 @@ public void scan(HttpMessage msg, String param, String origParamValue) {
...
 							return;
 						}
+					} catch (ArithmeticException ex){
+						if (this.debugEnabled) {
+							log.debug("Caught " + ex.getClass().getName() + " " + ex.getMessage() + "When performing integer math with the parameter value [" + origParamValue + "]");
+						}	
 					}
-				} catch (Exception e) {
-
+				} catch (Exception ex) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The rename is not really needed.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 27]
```diff
@@ -1401,6 +1360,85 @@ private boolean checkUnionErrors(RDBMS rdbms, HttpMessage msg, String response,
...
 	}
 
+	private void expressionBasedAttack(String param, String modifiedParamValue, String modifiedParamValueConfirm){
+		//those of you still paying attention will note that if handled as expressions (such as by a database), these represent the same value.
+		HttpMessage msg = getNewMsg();
+		setParameter(msg, param, modifiedParamValue);
+
+		try {
+			sendAndReceive(msg, false); //do not follow redirects
+		} catch (Exception ex) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The exception type was changed, why?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 28]
```diff
@@ -1401,6 +1360,85 @@ private boolean checkUnionErrors(RDBMS rdbms, HttpMessage msg, String response,
...
+				}
+				//confirm that a different parameter value generates different output, to minimise false positives
+				//this time param value will be different to original value and mismatch is expected in responses of original and this value
+				//Note that the two values are NOT equivalent, and the param value is different to the original
+				HttpMessage msgConfirm = getNewMsg();
+				setParameter(msgConfirm, param, modifiedParamValueConfirm);
+
+				try {
+					sendAndReceive(msgConfirm, false); //do not follow redirects
+				} catch (Exception ex) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Same here.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 29]
```diff
@@ -1451,6 +1489,38 @@ public static String getURLDecode(String msg) {
 		return result;
 	}
 
+	/**
+	* add two numbers with Arithmetic Overflow check
+	* @param firstNumber
+	* @param secondNumber
+	* @return
+	*/
+	public static int addWithOverflowCheck(int firstNumber, int secondNumber) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: - Should be private;
 - Could you add a TODO comment mentioning that this should be replaced with the corresponding `Math` method when targeting Java 8?

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java [hunk 30]
```diff
@@ -1451,6 +1489,38 @@ public static String getURLDecode(String msg) {
...
+	    return (int) result;
+	}
+
+	/**
+	* multiply two numbers with Arithmetic Overflow check
+	* @param firstNumber
+	* @param secondNumber
+	* @return
+	*/
+	public static int multiplyWithOverflowCheck(int firstNumber, int secondNumber) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Same comments as above.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java (hunk 27 above)
Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: I was getting this.
```
error: unreported exception HttpException; must be caught or declared to be thrown
    [javac] 			sendAndReceive(msg, false); //do not follow redirects
```
That rectify it.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java (hunk 19 above)
Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: The code looks a little nicer as other `false` initialisation were already there.

Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: OK.

File: src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java (hunk 27 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: Right, declaring that the method throws IOException allows to keep the previous behaviour.

Comment (rnehra01) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: So, should I remove HTTPException from
```
protected void sendAndReceive(HttpMessage msg) throws HttpException, IOException {
```
in `AbstractPlugin.java`

Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/TestSQLInjection.java]: That indeed can be removed (it already was), but I was referring to the method `expressionBasedAttack`, it is this one that should declare that throws IOException (to keep the previous behaviour).

**Comment by rnehra01** [comment]
@thc202 requested changes are done.

**Comment by thc202** [comment]
@rnehra01 thank you! How would you like to be credited?
https://github.com/zaproxy/zap-core-help/wiki/HelpCredits

**Comment by rnehra01** [comment]
@thc202 the way you like. And I will try to give as much time as possible to Zap(extension).

**Comment by rnehra01** [comment]
I think you are asking to put my name on the list **ZAP Extended Team**

**Comment by thc202** [comment]
Was asking which name/handle should be added :)

(For now it would be in that list.)

**Comment by rnehra01** [comment]
**rnehra01** (Ravinder Nehra)

**Comment by thc202** [comment]
Done in zaproxy/zap-core-help#67.
